common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -620,22 +698,29 @@ private synchronized KeyPair generateKeyPairFromAndroidKeyStore()
 
     /**
      * Read KeyPair from AndroidKeyStore.
+     *
+     * @return KeyPair. Null if there isn't any.
      */
-    private synchronized KeyPair readKeyPair() throws GeneralSecurityException, IOException {
+    @Nullable
+    private synchronized KeyPair readKeyPair() throws KeyStoreException {
         final String methodName = ":readKeyPair";
-        if (!doesKeyPairExist()) {
-            throw new KeyStoreException("KeyPair entry does not exist.");
-        }
-
         Logger.verbose(TAG + methodName, "Reading Key entry");
-        final KeyStore keyStore = KeyStore.getInstance(ANDROID_KEY_STORE);
-        keyStore.load(null);
 
-        final KeyStore.PrivateKeyEntry entry;
         try {
-            entry = (KeyStore.PrivateKeyEntry) keyStore.getEntry(
-                    KEY_STORE_CERT_ALIAS, null);
-        } catch (final RuntimeException e) {
+            final KeyStore keyStore = KeyStore.getInstance(ANDROID_KEY_STORE);
+            keyStore.load(null);
+
+            final Certificate cert = keyStore.getCertificate(KEY_STORE_CERT_ALIAS);
+            final Key privateKey = keyStore.getKey(KEY_STORE_CERT_ALIAS, null);
+
+            if (cert == null || privateKey == null) {
+                Logger.verbose(TAG + methodName, "Key entry doesn't exist.");
```

Comment:
>Logger [](start = 16, length = 6)

I want this to be in our telemetry stream.

Reply:
Telemetry is not wired into common yet. We could add this after Heidi has finished wiring her new telemetry work.

Reply:
bubbled up the exception